### DISCUSSION TITLE
Adding stream tag whenever `Source` changes frequency or sample rate.

### DISCUSSION THREAD

**Discussion by metasim** [FutureSDR/FutureSDR#166] [Q&A]
I have a use case where I need to (as precisely as possible) tag the I/Q stream coming from a `futuresdr::blocks::seify::Source` whenever the frequency or sample rate is changed via the remote API (whenever that happens I need to communicate to downstream `Block`s that the conditions under which I/Q is being sampled as changed).

1. Is there a recommended or sensible way of implementing this with the existing capabilities?
2. If not, is this something `futuresdr` should be extended to do (how exactly, TBD), or is this a niche enough use case I should just create my own custom "`seify::Source`"?

**Comment by vlovich** [comment]
@metasim did you ever find a solution here you liked?

I imagine you can just have a kernel that has a channel you enqueue into when it changes or even just a Arc<Mutex<(usize, Params)>> where the usize is a change counter the kernel remembers the last value and emits tags when it changes.

**Reply by metasim** [reply]
The approach I took was to create a custom source pass a tag, which required downstream kernels to pass tags along. It got the job done, but was a lot of boilerplate and hard to test. If I were to do it again I might use an Arc + Mutex, but I would have even less precision knowing where in the stream the change came. What I have kinda works, but feel janky.

**Comment by bastibl** [comment]
With Seify, I want to follow the semantics of Soapy. If you really care about the exact sample, you could first deactivate the RxStreamer and activate it again in the message handler. And also store some hint that a tag should be placed in the next call to work. (I didn't try, but that's what I would go for first.)

Then it's mainly an issue about forwarding tags through the flowgraph. I'm still not sure how this should be done. My current thought is that the blocks should take care of this internally (even though most upstream blocks don't do it, at the moment). If you have suggestions, let me know.

**Reply by vlovich** [reply]
I implemented a tag preservation kernel that wraps any other kernel that doesn't preserve (sample accurate). It's a bit of a pain but it works and I think it informs me of a better way to do it in FutureSDR natively.

Can paste the code a bit later for posterity (it's quite ugly), but the basic idea:

There's a "DelayedReader<Reader>" and "DelayedWriter<Writer>" that wrap the actual I/O objects so that they can intercept a kernel's call to produce/consume. When consume/produce is called, it just records the value. Then TagPreservingWrapper calls the inner kernel (which is given DelayedReader/delayedwriter as the types) for work and afterwards retrieves the consumed/produced counts and forwards the tags based on the decimation of consumed to produce before finally calling consumer/produce on the interior. There's also a "TagExtractable" trait which needs to be specialized per wrappable kernel to retrieve the DelayedReader/DelayedWriter from the wrapped kernel (since kernels don't have a trait to retrieve input/output)

It works but it's not ideal in that it doesn't work as cleanly as it could. Too much boilerplate and it leaks out (not a clean abstraction, requires teaching the system about each kernel I want to support)

I think the right way is for consume/produce to be called on WorkIO instead or something else injected into work so that kernels are more easily composable (eg io.transferred(reader, consume_n, writer, produce_n) which would just do reader.consume(n) / writer.produce(n) by default and capture that we're trying to forward samples from one to the other. WorkIo could have a completion: Box<dyn IoCompletion>.

Then TagPreserving's block just needs to swap out the completion object for its own before calling inner kernel's work. The custom one would then compute decimation and forward the tags to the correct writer samples before calling consumer/produce (and special case consume_n being the same as produce_n). This way there's no special DelayedReader / DelayedWriter / per kernel trait to access them. Indeed, you could even get rid of TagPreserving and just either make tag preservation the default or something you specify when adding the block / via the graph.

There's some open questions about how to make it possible to override the tag preservation when a kernel does it manually (in an explicit kernel it's the user's responsibility to know but with implicit default you could have the kernel interface indicate if it's a tag aware or agnostic kernel). Indeed with such an approach most kernels that forward tags today can be simplified and leverage the genetic machinery instead.

For StreamDuplicator/StreamDeinterleaver I think they'd need to handle the tags manually still. StreamSuplicator in theory could be handled by a fan_out call on IoCompletion if you want genetic forwarding, but I'm not sure it is generic. deinterleaver I'm also not sure if it's a generic forwarding primitive but it feels like it's not.

**Reply by bastibl** [reply]
Why don't you just forward the tags in the blocks?

https://github.com/FutureSDR/FutureSDR/blob/main/src/blocks/apply.rs#L102-L106
https://github.com/FutureSDR/FutureSDR/blob/main/src/blocks/fft.rs#L154-L157

**Reply by vlovich** [reply]
You could but none of the Fir blocks and a bunch of others don't do that. There's also a lot of duplicated tag forwarding logic, and with "NM" blocks (ApplyNM, Fir, etc) it's non trivial to map the tags.

**Reply by bastibl** [reply]
Forwarding tags is non-trivial. That's why it's not part of the runtime, but the blocks have to decide how the forwarding logic should be. Instead of the complicated wrapper, I would just extend existing blocks.
